content/blog/gsoc-25-netjsongraph.rst

```diff
@@ -0,0 +1,225 @@
+           overlay.add(new g.Circle(/* ... */));
+         }
+       }
+
+- **GUI Integration:** The sidebar was enhanced to recursively render nested data, allowing it to display detailed information for each client (e.g., MAC address).
+
+Feature: Zoom-Dependent Label Visibility
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- **The Problem:** In both map and graph views, displaying all node labels simultaneously at low zoom levels results in a cluttered and unreadable visualization.
+
+(Placeholder for GIF showing labels appearing/disappearing on zoom)
+
+- **The Mechanism & Logic:** I introduced ``showLabelsAtZoomLevel`` (for maps) and ``showGraphLabelsAtZoom`` (for graphs).
+
+- **Map Mode:** The solution uses the Leaflet ``zoomend`` event to dynamically update the ECharts series.
+
+    .. code-block:: javascript
+
+       // file: src/js/netjsongraph.render.js
+       self.leaflet.on("zoomend", () => {
+         const currentZoom = self.leaflet.getZoom();
+         const showLabel = currentZoom >= self.config.showLabelsAtZoomLevel;
+         self.echarts.setOption({
+           series: [{ id: "geo-map", label: { show: showLabel } }]
+         });
+       });
+
+- **Graph Mode:** The solution uses the ECharts ``graphRoam`` event and a custom label ``formatter`` function.
+
+    .. code-block:: javascript
+
+       // file: src/js/netjsongraph.render.js (inside generateGraphOption)
+       if (typeof self.config.showGraphLabelsAtZoom === "number") {
+         const threshold = self.config.showGraphLabelsAtZoom;
+         baseGraphLabel.formatter = (params) =>
+           getGraphZoom() >= threshold ? params.data.name || "" : "";
+       }
+
+Resiliency and Robustness Fixes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+- **Duplicate Node IDs:** I used a "defense-in-depth" strategy to prevent crashes from invalid data. The core logic is a new ``deduplicateNodesById`` utility.
+
+    .. code-block:: javascript
+
+       // file: src/js/netjsonWorker.js
+       deduplicateNodesById(nodes) {
+         const uniqueNodes = [];
+         const nodeIds = new Set();
+         nodes.forEach((node) => {
+           if (node.id) {
+             if (!nodeIds.has(node.id)) {
+               nodeIds.add(node.id);
+               uniqueNodes.push(node);
+             } else {
+               console.warn(`Duplicate node ID detected and skipped: ${node.id}`);
+             }
+           } else {
+             uniqueNodes.push(node); // Keep nodes without IDs
+           }
+         });
+         return uniqueNodes;
+       }
+
+- **Map Zoom Levels:** I corrected the map's behavior to respect tile provider limits and added CSS for visual feedback on the zoom controls.
+
+    .. code-block:: css
+
+       /* file: src/css/netjsongraph-theme.css */
+       .leaflet-control-zoom-in.leaflet-disabled,
+       .leaflet-control-zoom-out.leaflet-disabled {
+         opacity: 0.7;
+         cursor: default;
+       }
+
```

Comment:
Let's spend a few words about fixing the examples, changing the data of the examples to be human readable, updating the example images in the README.